// directus/dist/webhooks.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.unregister = exports.register = void 0;
const axios_1 = __importDefault(require("axios"));
const database_1 = __importDefault(require("./database"));
const emitter_1 = __importDefault(require("./emitter"));
const logger_1 = __importDefault(require("./logger"));
const lodash_1 = require("lodash");
let registered = [];
async function register() {
    unregister();
    const database = (0, database_1.default)();
    const webhooks = await database.select('*').from('directus_webhooks').where({ status: 'active' });
    for (const webhook of webhooks) {
        if (webhook.actions === '*') {
            const event = 'items.*';
            const handler = createHandler(webhook);
            emitter_1.default.on(event, handler);
            registered.push({ event, handler });
        }
        else {
            for (const action of webhook.actions.split(',')) {
                const event = `items.${action}`;
                const handler = createHandler(webhook);
                emitter_1.default.on(event, handler);
                registered.push({ event, handler });
            }
        }
    }
}
exports.register = register;
function unregister() {
    for (const { event, handler } of registered) {
        emitter_1.default.off(event, handler);
    }
    registered = [];
}
exports.unregister = unregister;
function createHandler(webhook) {
    return async (data) => {
        const collectionAllowList = webhook.collections.split(',');
        if (collectionAllowList.includes('*') === false && collectionAllowList.includes(data.collection) === false)
            return;
        const webhookPayload = (0, lodash_1.pick)(data, [
            'event',
            'accountability.user',
            'accountability.role',
            'collection',
            'item',
            'action',
            'payload',
        ]);
        try {
            await (0, axios_1.default)({
                url: webhook.url,
                method: webhook.method,
                data: webhook.data ? webhookPayload : null,
            });
        }
        catch (error) {
            logger_1.default.warn(`Webhook "${webhook.name}" (id: ${webhook.id}) failed`);
            logger_1.default.warn(error);
        }
    };
}
